tensorwaves/interfaces.py

```diff
@@ -0,0 +1,102 @@
+"""Defines top level interfaces of tensorwaves.
+
+"""
+
+from abc import ABC, abstractmethod
+
+
+class Function(ABC):
+    """
+    Defines the interface of a callable function.
+    The parameters of the model are separated from the domain variables. This
+    follows the mathematical definition, in which a function defines its domain
+    and parameters. However specific points in the domain are not relevant.
+    Hence while the domain variables are the argument of the evaluation
+    (see :func:`~Function.__call__`), the parameters are controlled via a
+    getter and setter (see :func:`~Function.parameters`). The reason for this
+    separation is to facilitate the events when parameters have changed.
+    """
+    @abstractmethod
+    def __call__(self, dataset: dict) -> list:
+        """Evaluate the function.
+
+        Args:
+            dataset: a dictionary with domain variable names as keys and a
+            an list or `numpy.array` of  with a
+
+        Return:
+            list or :class:`numpy.array` of values
+        """
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        """
+        """
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        """
+        """
+
+
+class Estimator(ABC):
+    @abstractmethod
+    def __call__(self) -> float:
+        pass
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        pass
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        pass
+
+
+class Kinematics(ABC):
+    @abstractmethod
+    def convert(self, events: dict) -> dict:
+        pass
+
+    @abstractmethod
+    def reduce_to_phase_space(self, events: dict) -> dict:
```

Comment:
Is this necessary? Or does a `is_within_phase_space` suffice?

Reply:
I like `is_within_phase_space` better

Reply:
So the following?
```python
def is_within_phase_space(self, event) -> bool:
    """Check whether event lies within phase space."""
```
Or maybe this?
```python
def is_within_phase_space(self, events: dict) -> list:
    """Check which events lie within phase space."""
```

Reply:
Definitely the second one. Because the first thing we had in the beginning and this event based way of processing did not play together well with the vectorized forms in the rest of the code

Reply:
https://github.com/ComPWA/tensorwaves/pull/30/commits/d55d6b53324828ea50f359b2dae0870fbdae42ea